### PR TITLE
support fp16 shgemm under openblas

### PR DESCRIPTION
**Pull request by langc23**
# Purpose
This PR is to support fp16 shgemm under openblas. We conducted tests using vLLM on the following platform. 
With using this patch, vLLM demonstrates faster inference speed under fp16.

**Platform info:**
Architecture:             riscv64
  Byte Order:             Little Endian
CPU(s):                   64
  On-line CPU(s) list:    0-63
Vendor ID:                0x5b7
  BIOS Vendor ID:         SOPHGO
  Model name:             -
    BIOS Model name:      SG2044 Not Set CPU @ 2.6GHz
    BIOS CPU family:      513
    CPU family:           0x80000000090c0d00
    Model:                0x2047000
    Thread(s) per core:   1
    Core(s) per socket:   64
    Socket(s):            1
    Frequency boost:      disabled
    CPU(s) scaling MHz:   100%
    CPU max MHz:          2600.0000
    CPU min MHz:          1000.0000
Caches (sum of all):      
  L1d:                    4 MiB (64 instances)
  L1i:                    4 MiB (64 instances)
  L2:                     32 MiB (16 instances)
  L3:                     64 MiB (1 instance)
Vulnerabilities:          
  Gather data sampling:   Not affected
  Itlb multihit:          Not affected
  L1tf:                   Not affected
  Mds:                    Not affected
  Meltdown:               Not affected
  Mmio stale data:        Not affected
  Reg file data sampling: Not affected
  Retbleed:               Not affected

ISA: rv64imafdcv_zicbom_zicboz_zicntr_zicond_zicsr_zifencei_zihintntl_zihintpause_zihpm_zawrs_zfa_zfh_zfhmin_zca_zcb_zcd_zba_zbb_zbc_zbs_zve32f_zve32x_zve64d_zve64f_zve64x_zvfh_zvfhmin_sscofpmf_sstc_svinval_svnapot_svpbmt

**Branch**
openblas: develop
torch: develop
vllm: main

# Test Plan
Base: without this PR
Pytorch use OpenBLAS FP16 GEMM: use this PR

**Base**
export VLLM_CPU_OMP_THREADS_BIND=0-63
export VLLM_CPU_KVCACHE_SPACE=60

vllm bench latency \
  --model /home/models/Qwen2.5-7B-Instruct \
  --tensor-parallel-size 1\
  --dtype float16 \
  --input-len 16 \
  --output-len 16 \
  --enforce-eager \
  --max-model-len 8192 \
  --max-num-batched-tokens 8192 \
  --batch-size 1 \
  --n 1 \
  --num-iters-warmup 5 \
  --num-iters 8 \
  --seed 42 \
  --output-json ./latency_results_fp16_latency_base.json

**Pytorch use OpenBLAS FP16 GEMM**
export VLLM_CPU_OMP_THREADS_BIND=0-63
export VLLM_CPU_KVCACHE_SPACE=60

vllm bench latency \
  --model /home/models/Qwen2.5-7B-Instruct \
  --tensor-parallel-size 1\
  --dtype float16 \
  --input-len 16 \
  --output-len 16 \
  --enforce-eager \
  --max-model-len 8192 \
  --max-num-batched-tokens 8192 \
  --batch-size 1 \
  --n 1 \
  --num-iters-warmup 5 \
  --num-iters 8 \
  --seed 42 \
  --output-json ./latency_results_fp16_latency_with_openblas_support.json

# Result
**Base**
{
    "avg_latency": 62.53946338250171,
    "latencies": [
        58.46783778001554,
        58.230652199999895,
        58.335780619992875,
        59.77051957999356,
        58.587668860011036,
        59.31567866000114,
        58.460076240007766,
        89.14749311999185
    ],
    "percentiles": {
        "10": 58.30424209399498,
        "25": 58.42900233500404,
        "50": 58.52775332001329,
        "75": 59.429388889999245,
        "90": 68.58361164199304,
        "99": 87.09110497219196
    }
}

**Pytorch use OpenBLAS FP16 GEMM**
{
    "avg_latency": 32.42863222499727,
    "latencies": [
        30.742418120033108,
        33.67000828002347,
        29.747197599965148,
        32.11275753995869,
        34.566938299976755,
        30.849812360014766,
        34.46360486000776,
        33.27632073999848
    ],
    "percentiles": {
        "10": 30.44385196401272,
        "25": 30.82296380001935,
        "50": 32.69453913997859,
        "75": 33.86840742501954,
        "90": 34.49460489199846,
        "99": 34.55970495917892
    }
}

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01

### CODE REVIEW

**Comment by langc23** [comment]
@malfet @Skylion007 @albanD @EikanWang @jananisriram @PaulZhang12  Hello, this PR is to support fp16 shgemm under openblas, could someone review the codes? Thanks.

**Review by aditew01 — APPROVED** [review]

**Comment by langc23** [comment]
> Thanks! Looks good to me, can you please post benchmarks with `torch.nn.Linear` or `scaled_dot_product_attention` in description for understanding the native speedups.

@aditew01 Thanks for your reviewing. We have supplemented the tests as shown below.

**FP16 CPU performance improvement on riscv64 (SG2044)**
We benchmarked FP16 nn.Linear and scaled_dot_product_attention on a riscv64 SG2044 system with 64 CPU threads, using OpenBLAS FP16 GEMM (shgemm) enabled by this PR.

**Environment**
Hardware:    riscv64 SG2044 (64 CPU cores)
BLAS:        OpenBLAS 0.3.30.dev (RISCV64_ZVL128B, OpenMP, FP16 GEMM enabled)
PyTorch:     built from source, CPU only
Dtype:       torch.float16
Device:      cpu
Threads:     OMP_NUM_THREADS=64, OPENBLAS_NUM_THREADS=64, torch.set_num_threads(64)

**Runtime environment variables**
export OMP_NUM_THREADS=64
export OPENBLAS_NUM_THREADS=64

**Scripts**
linear_bench_fp16.py
```
#!/usr/bin/env python3
import time
import argparse
import torch
import torch.nn as nn


def bench_linear(M, K, N, dtype=torch.float16,
                 iters_warmup=3, iters=10):
    x = torch.randn(M, K, dtype=dtype, device="cpu")
    linear = nn.Linear(K, N, bias=False, dtype=dtype, device="cpu")

    # warmup
    with torch.no_grad():
        for _ in range(iters_warmup):
            _ = linear(x)

    start = time.perf_counter()
    with torch.no_grad():
        for _ in range(iters):
            _ = linear(x)
    end = time.perf_counter()

    avg_ms = (end - start) * 1000.0 / iters
    # 2 * M * K * N FLOPs for GEMM
    flops = 2.0 * M * K * N
    gflops = flops / (avg_ms / 1e3) / 1e9
    return avg_ms, gflops


def main():
    parser = argparse.ArgumentParser(
        description="FP16 nn.Linear CPU benchmark (OpenBLAS shgemm)"
    )
    parser.add_argument("--threads", type=int, default=64,
                        help="torch.set_num_threads")
    parser.add_argument("--warmup", type=int, default=3)
    parser.add_argument("--iters", type=int, default=10)
    args = parser.parse_args()

    torch.set_num_threads(args.threads)

    print("=== nn.Linear FP16 CPU benchmark ===")
    print(f"threads={args.threads}, warmup={args.warmup}, iters={args.iters}\n")

    # transformer-like linear shapes
    shapes = [
        (128, 4096, 4096),
        (128, 4096, 11008),
        (128, 11008, 4096),
        (256, 4096, 4096),
    ]

    for (M, K, N) in shapes:
        avg_ms, gflops = bench_linear(
            M, K, N,
            dtype=torch.float16,
            iters_warmup=args.warmup,
            iters=args.iters,
        )
        print(
            f"M={M}, K={K}, N={N}: "
            f"avg_latency={avg_ms:.3f} ms, throughput={gflops:.1f} GFLOPS"
        )


if __name__ == "__main__":
    main()
```
sdpa_bench_fp16.py
```
#!/usr/bin/env python3
import time
import argparse
import torch
import torch.nn.functional as F


def bench_sdpa(B, H, S, D, dtype=torch.float16,
               causal=True, iters_warmup=3, iters=5):
    q = torch.randn(B, H, S, D, dtype=dtype, device="cpu")
    k = torch.randn(B, H, S, D, dtype=dtype, device="cpu")
    v = torch.randn(B, H, S, D, dtype=dtype, device="cpu")

    # warmup
    with torch.no_grad():
        for _ in range(iters_warmup):
            _ = F.scaled_dot_product_attention(
                q, k, v, is_causal=causal
            )

    start = time.perf_counter()
    with torch.no_grad():
        for _ in range(iters):
            _ = F.scaled_dot_product_attention(
                q, k, v, is_causal=causal
            )
    end = time.perf_counter()

    avg_ms = (end - start) * 1000.0 / iters

    # rough FLOPs estimate: O(B * H * S^2 * D)
    flops = 2.0 * B * H * (S * S) * D
    gflops = flops / (avg_ms / 1e3) / 1e9
    return avg_ms, gflops


def main():
    parser = argparse.ArgumentParser(
        description="FP16 scaled_dot_product_attention CPU benchmark"
    )
    parser.add_argument("--threads", type=int, default=64,
                        help="torch.set_num_threads")
    parser.add_argument("--warmup", type=int, default=3)
    parser.add_argument("--iters", type=int, default=5)
    parser.add_argument("--no_causal", action="store_true",
                        help="disable causal mask")
    args = parser.parse_args()

    torch.set_num_threads(args.threads)
    causal = not args.no_causal

    print("=== scaled_dot_product_attention FP16 CPU benchmark ===")
    print(
        f"threads={args.threads}, warmup={args.warmup}, iters={args.iters}, "
        f"causal={causal}\n"
    )

    # transformer-like SDPA shapes
    shapes = [
        (1, 32, 128, 128),
        (1, 32, 512, 128),
    ]

    for (B, H, S, D) in shapes:
        avg_ms, gflops = bench_sdpa(
            B, H, S, D,
            dtype=torch.float16,
            causal=causal,
            iters_warmup=args.warmup,
            iters=args.iters,
        )
        print(
            f"B={B}, H={H}, S={S}, D={D}: "
            f"avg_latency={avg_ms:.3f} ms, throughput={gflops:.1f} GFLOPS"
        )


if __name__ == "__main__":
    main()
```

**Commands**
without this PR
```
# FP16 nn.Linear
python3 linear_bench_fp16.py \
  --threads 64 --warmup 3 --iters 10 \
  > linear_base_64t.txt

# FP16 scaled_dot_product_attention
python3 sdpa_bench_fp16.py \
  --threads 64 --warmup 3 --iters 5 \
  > sdpa_base_64t.txt
```
with this PR
```
# FP16 nn.Linear
python3 linear_bench_fp16.py \
  --threads 64 --warmup 3 --iters 10 \
  > linear_pr_64t.txt

# FP16 scaled_dot_product_attention
python3 sdpa_bench_fp16.py \
  --threads 64 --warmup 3 --iters 5 \
  > sdpa_pr_64t.txt
```
**Results**
nn.Linear FP16 (64 threads)
This PR improves FP16 GEMM throughput significantly:
**20× – 45×** speedup on typical transformer shapes.
```
Shape (M,K,N)   | Baseline (ms) | This PR (ms) | Speedup
128,4096,4096   | 424.844 | 18.853  | 22.5×
128,4096,11008 | 1135.443 | 57.588  | 19.7×
128,11008,4096 | 1290.106 | 28.253  | 45.6×
256,4096,4096   | 915.745 | 24.597  | 37.2×
```
scaled_dot_product_attention FP16 (64 threads)
```
(B,H,S,D)  | Baseline (ms) | This PR (ms) | Speedup
1,32,128,128 | 17.680  | 15.078 | 1.17×
1,32,512,128 | 439.371 | 83.532 | 5.26×
```
**Summary**
On this setup, enabling OpenBLAS FP16 GEMM (shgemm) via this PR improves FP16 nn.Linear throughput by ~20–45× on typical transformer shapes, and FP16 scaled_dot_product_attention by up to ~5.3× for long sequence length (S=512) with 64 CPU threads.

**Comment by langc23** [comment]
@aditew01 Hi, we hope the PR could merged into the mainline.  Could you please take a time review the results and enable the CI? Thanks.

**Comment by langc23** [comment]
@aditew01 Could you please rebuild the error check, it appears that this was not caused by this PR.

```
+ echo 'The next three invocations are expected to fail with invalid command error messages.'
The next three invocations are expected to fail with invalid command error messages.
+ get_exit_code python setup.py bad_argument
+ set +e
+ python setup.py bad_argument
Building wheel torch-2.10.0a0+gitc2d423b
Found cmake (/usr/local/bin/cmake) version: 4.0.0 (>=3.27)
invalid command 'bad_argument'
+ retcode=1
+ set -e
+ return 1
+ get_exit_code python setup.py 'clean]'
+ set +e
+ python setup.py 'clean]'
Building wheel torch-2.10.0a0+gitc2d423b
Found cmake (/usr/local/bin/cmake) version: 4.0.0 (>=3.27)
invalid command name 'clean]'
+ retcode=1
+ set -e
+ return 1
+ get_exit_code python setup.py clean bad_argument
+ set +e
+ python setup.py clean bad_argument
Building wheel torch-2.10.0a0+gitc2d423b
Found cmake (/usr/local/bin/cmake) version: 4.0.0 (>=3.27)
invalid command 'bad_argument'
+ retcode=1
+ set -e
+ return 1
+ [[ linux-noble-riscv64-py3.12-gcc14 != *libtorch* ]]
+ [[ linux-noble-riscv64-py3.12-gcc14 != *rocm* ]]
+ [[ linux-noble-riscv64-py3.12-gcc14 != *xla* ]]
+ [[ linux-noble-riscv64-py3.12-gcc14 != *riscv64* ]]
+ python setup.py clean
Building wheel torch-2.10.0a0+gitc2d423b
Found cmake (/usr/local/bin/cmake) version: 4.0.0 (>=3.27)
/opt/riscv-cross-env/cross/lib/python3.12/site-packages/setuptools/config/_apply_pyprojecttoml.py:82: SetuptoolsDeprecationWarning: `project.license` as a TOML table is deprecated
!!

        ********************************************************************************
        Please use a simple string containing a SPDX expression for `project.license`. You can also use `project.license-files`. (Both options available on setuptools>=77.0.0).

        By 2026-Feb-18, you need to update your project and remove deprecated calls
        or your builds will no longer be supported.

        See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
        ********************************************************************************

!!
  corresp(dist, value, root_dir)
running clean
+ [[ linux-noble-riscv64-py3.12-gcc14 == *xla* ]]
+ python -m build --wheel --no-isolation
/opt/riscv-cross-env/cross/bin/python3: No module named build
+ ccache_epilogue
+ ccache --show-stats
Local storage:
  Cache size (GiB): 0.0 / 5.0 ( 0.00%)
Error: Process completed with exit code 1.
```

**Review by albanD — APPROVED** [review]
Sounds fair!

**Comment by langc23** [comment]
> Would you be able to rebase this PR on the latest main so we can have a clean CI run?你能将这个 PR 重新基于最新的 main 分支进行合并，以便我们有一个干净的 CI 运行吗？

@albanD  Hi, Ihave rebase this PR on the latest main. Plaese help rebuild the CI. Thanks :)
